perun/collect/trace/strategy.py

```diff
@@ -28,213 +28,287 @@ class _Status(IntEnum):
-
-    seen = set()
-    unique = []
-    # Prioritize paired rules - we can't afford to remove paired rule instead of a single rule
-    # Also prioritize sampled rules
-    for paired_rules in [paired_sampled, paired]:
-        for probe in paired_rules:
-            if probe['name'] not in seen and probe['pair'] not in seen:
-                # Add new unique rule
-                seen.add(probe['name'])
-                seen.add(probe['pair'])
-                unique.append(probe)
-
-    # Now add single rules that are not duplicate
-    for single_rules in [single_sampled, single]:
-        for probe in single_rules:
-            if probe['name'] not in seen:
-                # Add new unique rule
-                seen.add(probe['name'])
-                unique.append(probe)
-
-    return unique
-
-
-def _merge_extracted_with_custom(extracted, custom):
-    """Merges extracted probe locations with user-specified ones, where the extracted have
-    lower priority and can be overwritten by the user supplied rules.
-
-    :param list extracted: list of extracted probes represented as a dictionaries
-    :param list custom: list of user-specified probes represented also as a dictionaries
-    :return list: the resulting list of merged probe locations
+            basic[k] = v
+
+    # Process the basic probes - find endings and compare them
+    beginnings, endings = dict(), dict()
+    for k, v in basic.items():
+        # Classify the probe as starting or ending according to the known suffixes
+        if not _check_suffix(v, suffix_pairs.values(), endings):
+            if not _check_suffix(v, suffix_pairs.keys(), beginnings):
+                # Not starting nor ending suffix, treat as regular probe name
+                beginnings.setdefault(v['name'], {'sample': v['sample'], 'suffix': []})
+    return delimited, beginnings, endings
+
+
+def _check_suffix(probe, suffix_group, cls):
+    """Checks if the probe name contains any suffix from the suffix group and if yes, classifies it
+    into the specified cls. Also stores the delimiter between the probe name and suffix, if any.
+
+    :param dict probe: the given probe specification
+    :param collection suffix_group: the collection of suffixes to check
+    :param dict cls: the resulting class
+    """
+    default = {'sample': probe['sample'], 'suffix': []}
+    # Iterate the suffixes from suffix group and check if the probe name ends with one of them
+    for suffix in suffix_group:
+        if probe['name'].lower().endswith(suffix):
+            name = probe['name'][:-len(suffix)]
+            # Store the delimiter if any
+            delimiter = ''
+            while name[-1] in _SUFFIX_DELIMITERS:
+                delimiter = name[-1] + delimiter
+                name = name[:-1]
+            cls.setdefault(name, default)['suffix'].append((suffix, delimiter))
+            return True
+    return False
+
+
+def _pair_suffixes(b_suffixes, e_suffixes, pairing):
```

Comment:
Suggestion: the "pairing" parameter sounds strange, since it is function. It is callable at several places, so I would suggest  either to rename it to "pairing_func" or something doable, e.g., "pair_by"